specs/deposit_gas_metering.md

```diff
@@ -0,0 +1,150 @@
+
+### Gas Stipend
+
+Can also provide a stipend to the guaranteed gas of every transaction.
+The stipend is bought with gas that was spent executing the deposit logic;
+however, the stipend still needs to be included in the sum of bought
+guaranteed gas.
+
+TODO: How much / do we actually need this?
+
+If a gas stipend is provided, we only buy the amount of guaranteed gas
+in excess of the gas stipend.
+
+### Additional Gas
+
+Users can request additional gas for their L2 transaction. This gas is
+refundable is less than the gas limit is used. However, if there is not
+enough available gas on L2 or the gas price not high enough, the additional
+gas will not be provided on L2.
+
+Additional gas is bought with the following:
+
+- `gasPrice`: L2 gas price to buy.
+- `additionalGas`: Amount of additional gas requested
+
+The additional gas will be bought with the `from` account's balance. We do
+not require that the user mint enough L2 eth to cover the additional gas fee,
+but do suggest that tooling checks that the user has enough balance for it.
+
+### Limiting Guaranteed Gas
+
+The deposit feed contract must limit the total amount of guaranteed gas in a
+single L1 block. This is to limit the amount of gas that is used on L2.
+
+This is done with a hard limit on the amount of guaranteed gas that can be
+bought. To reduce PGAs if the deposit mechanism is congested, we also implement
+an EIP-1559 style fee market with the following pseudo code:
+
+```text
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+
+curr_basefee: u128, curr_num: u64, curr_bought_gas: u64 = load_and_unpack_storage()
+gas_target_limit: u64, gas_sanity_limit: u64 = load_and_unpack_storage2()
+
+# // implies floor division, however because gas_delta is always positive, it is the same as truncating (aka round to 0) division
+# If first deposit of this block, calculate the new basefee and store other info as well.
+if curr_num != block.number {
+    if curr_bought_gas == gas_target {
+        new_basefee := curr_basefee
+    } else if curr_bought_gas > gas_target {
+        gas_delta     := curr_bought_gas - gas_target
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        basefee_delta := max(basefee_delta, 1) # TODO: Why does 1559 have this asymmetry?
+        new_basefee   := curr_basefee + basefee_delta
+    } else {
+        gas_delta     := gas_target - curr_bought_gas
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        // Fun fact, geth doesn't let the new_basefee get below 0.
+        new_basefee   := curr_basefee - basefee_delta
+    }
+    curr_basefee := new_basefee
+    curr_number := block.number
+    curr_bought_gas := 0
+    pack_and_store(curr_basefee, curr_number, curr_bought_gas)
+
+}
+
+curr_bought_gas += required_gas
+require(curr_bought_gas <= GUARANTEED_GAS_LIMIT)
+require(curr_bought_gas <= SANITY_GAS_LIMIT)
+gas_cost = requested_gas * curr_basefee
```

Comment:
This looks different than the original pseudocode I wrote. The pack_and_store should be at the end, and always run. Otherwise the bought_gas never increases between the deposits in the same block.

Reply:
Nice catch. I did update it try to figure out wtf eip 1559 is doing with the basefee calc.